File: .tmp_py4013/batch_08_dump/2920_maximum_points_after_collecting_coins_from_all_nodes.js
// LeetCode 2920 - Maximum Points After Collecting Coins From All Nodes
// https://leetcode.com/problems/maximum-points-after-collecting-coins-from-all-nodes/

/**
 * @param {number[][]} edges
 * @param {number[]} coins
 * @param {number} k
 * @return {number}
 */
var maximumPoints = function(edges, coins, k) {
    const n = coins.length;
    const g = Array.from({ length: n }, () => []);
    for (const [a, b] of edges) {
        g[a].push(b);
        g[b].push(a);
    }
    const memo = new Map();
    const dfs = (u, p, shifts) => {
        if (shifts > 14) shifts = 14;
        const key = (u << 5) | shifts;
        if (memo.has(key)) return memo.get(key);
        const c = coins[u] >> shifts;
        let opt1 = c - k, opt2 = Math.floor(c / 2);
        for (const v of g[u]) {
            if (v === p) continue;
            opt1 += dfs(v, u, shifts);
            opt2 += dfs(v, u, shifts + 1);
        }
        const best = Math.max(opt1, opt2);
        memo.set(key, best);
        return best;
    };
    return dfs(0, -1, 0);
};

File: .tmp_rs4013/rust_check_12/2920_maximum_points_after_collecting_coins_from_all_nodes.rs
struct Solution;
// LeetCode 2920 - Maximum Points After Collecting Coins From All Nodes
// https://leetcode.com/problems/maximum-points-after-collecting-coins-from-all-nodes/

use std::collections::HashMap;

impl Solution {
    pub fn maximum_points(edges: Vec<Vec<i32>>, coins: Vec<i32>, k: i32) -> i32 {
        let n = coins.len();
        let mut g = vec![Vec::new(); n];
        for e in edges {
            g[e[0] as usize].push(e[1] as usize);
            g[e[1] as usize].push(e[0] as usize);
        }
        fn dfs(
            u: usize,
            p: i32,
            mut shifts: i32,
            g: &[Vec<usize>],
            coins: &[i32],
            k: i32,
            memo: &mut HashMap<(usize, i32), i32>,
        ) -> i32 {
            if shifts > 14 {
                shifts = 14;
            }
            if let Some(&v) = memo.get(&(u, shifts)) {
                return v;
            }
            let c = coins[u] >> shifts;
            let mut opt1 = c - k;
            let mut opt2 = c / 2;
            for &v in &g[u] {
                if v as i32 == p {
                    continue;
                }
                opt1 += dfs(v, u as i32, shifts, g, coins, k, memo);
                opt2 += dfs(v, u as i32, shifts + 1, g, coins, k, memo);
            }
            let best = opt1.max(opt2);
            memo.insert((u, shifts), best);
            best
        }
        let mut memo = HashMap::new();
        dfs(0, -1, 0, &g, &coins, k, &mut memo)
    }
}

fn main() {}
